splinterd/src/routes/circuit.rs

```diff
@@ -0,0 +1,531 @@
+use splinter::futures::{future::IntoFuture, Future};
+use splinter::rest_api::{Method, Resource, RestResourceProvider};
+
+use std::collections::HashMap;
+use std::error::Error as StdError;
+use std::sync::{Arc, RwLock};
+
+#[derive(Debug)]
+pub enum CircuitRouteError {
+    BuildError(CircuitBuildError),
+    NotFound(String),
+}
+
+impl StdError for CircuitRouteError {
+    fn source(&self) -> Option<&(dyn StdError + 'static)> {
+        match self {
+            CircuitRouteError::BuildError(err) => Some(err),
+            CircuitRouteError::NotFound(_) => None,
+        }
+    }
+}
+
+impl std::fmt::Display for CircuitRouteError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            CircuitRouteError::BuildError(err) => write!(f, "Unable to build circuit: {}", err),
+            CircuitRouteError::NotFound(msg) => write!(f, "Circuit not found: {}", msg),
+        }
+    }
+}
+
+impl From<CircuitBuildError> for CircuitRouteError {
+    fn from(err: CircuitBuildError) -> Self {
+        CircuitRouteError::BuildError(err)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct CircuitResponse {
+    id: String,
+    auth: AuthorizationType,
+    members: Vec<String>,
+    roster: Roster,
+    persistence: PersistenceType,
+    durability: DurabilityType,
+    routes: RouteType,
+    circuit_management_type: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct ListCircuitsResponse {
+    data: Vec<CircuitResponse>,
+    paging: Paging,
+}
+
+#[derive(Clone)]
+pub struct CircuitResourceProvider {
+    node_id: String,
+    state: Arc<RwLock<SplinterState>>,
+}
+
+impl CircuitResourceProvider {
+    pub fn new(node_id: String, state: Arc<RwLock<SplinterState>>) -> Self {
+        Self { node_id, state }
+    }
+}
+
+impl RestResourceProvider for CircuitResourceProvider {
+    fn resources(&self) -> Vec<Resource> {
+        vec![
+            make_fetch_circuit_resource(self.state.clone()),
+            make_list_circuits_resource(self.state.clone()),
+        ]
+    }
+}
+
+fn make_fetch_circuit_resource(state: Arc<RwLock<SplinterState>>) -> Resource {
+    Resource::new(Method::Get, "/circuits/{circuit_id}", move |r, _| {
+        fetch_circuit(r, web::Data::new(state.clone()))
+    })
+}
+
+fn make_list_circuits_resource(state: Arc<RwLock<SplinterState>>) -> Resource {
+    Resource::new(Method::Get, "/circuits", move |r, _| {
+        list_circuits(r, web::Data::new(state.clone()))
+    })
+}
+
+pub fn fetch_circuit(
+    request: HttpRequest,
+    state: web::Data<Arc<RwLock<SplinterState>>>,
+) -> Box<dyn Future<Item = HttpResponse, Error = Error>> {
+    let circuit_id = request
+        .match_info()
+        .get("circuit_id")
+        .unwrap_or("")
+        .to_string();
+    Box::new(
+        web::block(move || {
+            let circuit = {
+                let state = state.read().expect("Splinter state lock was poisoned");
```

Comment:
Shouldn't this return a 500 internal server error rather than panic?

Reply:
yep good catch

Reply:
fixed